### DISCUSSION TITLE
overriding generateId does nothing

### DISCUSSION THREAD

**Discussion by XiXiongMaoXiong** [socketio/socket.io#4190] [Q&A] [answered]
I've tried all of the 3 methods shown in the picture (method 2 commented out atm), the 2 "official" ones don't work and by overriding it I get nothing even though the code inside the generateId gets called (I tested it by console logging), but my custom ID is ignored, as you can see in the console log that the generated ID is not what I expected.

First way, overriding generateId by passing a callback:
![image](https://user-images.githubusercontent.com/23288867/143774008-db258563-60db-44b8-917b-546aa0370aba.png)

Second way, overriding generateId by not passing a callback:
![image](https://user-images.githubusercontent.com/23288867/143774232-5934f849-1040-4189-8b8a-e6c802494db0.png)

Third way, by directly modifying the the id property on the socket object:
![image](https://user-images.githubusercontent.com/23288867/143774433-5da3fd2f-788f-4044-b989-7b78d345dd2b.png)

As you can see in socket.io types, the id is declared as readonly:
![image](https://user-images.githubusercontent.com/23288867/143774486-22a832f7-c83e-46e7-9fed-b99f9c9bdb22.png)

I couldn't find any useful answers in SO, neither the SocketIO docs are saying anything about this.
Not to forget mentioning that back when I was using v3 of SIO there was not warning about modifying the socket.id as the type was not declared as readonly.

Please give me a proper solution, or should I continue doing this? And if you're "security concerns" are about the uniqueness I can surely tell you the nanoid package never fails.

**Comment by darrachequesne** [comment, marked as answer]
The `generateId` method applies to the session ID of the low-level connection (the `sid` in the query parameters).

You can indeed override the Socket.IO id in a middleware before the connection:

```js
io.use((socket, next) => {
  socket.id = "...";
  next();
});
```

This isn't really documented because it's not a frequent use case, could you please explain yours?
